extensions/s3/src/main/java/io/deephaven/extensions/s3/S3ChannelContext.java

```diff
@@ -0,0 +1,430 @@
+
+        @Override
+        public void accept(ByteBuffer byteBuffer, Throwable throwable) {
+            completedAt = Instant.now();
+            if (log.isDebugEnabled()) {
+                if (byteBuffer != null) {
+                    log.debug("send complete: {} {}", requestStr(), Duration.between(createdAt, completedAt));
+                } else {
+                    log.debug("send error: {} {}", requestStr(), Duration.between(createdAt, completedAt));
+                }
+            }
+        }
+
+        // --------------------------------------------------------------------------------------------------
+
+        @Override
+        public CompletableFuture<ByteBuffer> prepare() {
+            final CompletableFuture<ByteBuffer> future = new CompletableFuture<>();
+            producerFuture = future;
+            return future;
+        }
+
+        @Override
+        public void onResponse(GetObjectResponse response) {
+            this.response = response;
+        }
+
+        @Override
+        public void onStream(SdkPublisher<ByteBuffer> publisher) {
+            publisher.subscribe(new Request.Sub());
+        }
+
+        @Override
+        public void exceptionOccurred(Throwable error) {
+            producerFuture.completeExceptionally(error);
+        }
+
+        // --------------------------------------------------------------------------------------------------
+
+        private ByteBuffer get() throws ExecutionException, InterruptedException, TimeoutException {
+            if (released.isDone()) {
+                throw new IllegalStateException(String.format("Trying to get data after release, %s", requestStr()));
+            }
+            // Giving our own get() a bit of overhead - the clients should already be constructed with appropriate
+            // apiCallTimeout.
+            final long readNanos = instructions.readTimeout().plusMillis(100).toNanos();
+            return consumerFuture.get(readNanos, TimeUnit.NANOSECONDS);
+        }
+
+        private boolean isFragment(long fragmentIndex) {
+            return this.fragmentIndex == fragmentIndex;
+        }
+
+        private int requestLength() {
+            return (int) (to - from + 1);
+        }
+
+        private GetObjectRequest getObjectRequest() {
+            return GetObjectRequest.builder()
+                    .bucket(uri.bucket().orElseThrow())
+                    .key(uri.key().orElseThrow())
+                    .range("bytes=" + from + "-" + to)
+                    .build();
+        }
+
+        private String requestStr() {
+            if (uri != null) {
+                return String.format("ctx=%d ix=%d [%d, %d]/%d %s/%s", System.identityHashCode(S3ChannelContext.this),
+                        fragmentIndex, from, to, requestLength(), uri.bucket().orElseThrow(), uri.key().orElseThrow());
+            } else {
+                return String.format("ctx=%d ix=%d [%d, %d]/%d", System.identityHashCode(S3ChannelContext.this),
```

Comment:
Do we want to allow requests to exist with a `null` URI? I think the request should hold the URI.